### PR TITLE
Add filters for species, tissue, treatment and outcome on BioCollection

### PR DESCRIPTION
**Pull request by plojyon**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Amazon ECR.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/filters.py [hunk 1]
```diff
@@ -32,6 +34,56 @@ class BioCollectionFilter(CollectionFilter):
...
+
+    def _filter_by_outcome(self, queryset, name, value):
+        locations = [
+            "descriptor__response_and_survival_analysis__clinical_benefit",
+            "descriptor__response_and_survival_analysis__confirmed_bor",
+            "descriptor__response_and_survival_analysis__unconfirmed_bor",
+            "descriptor__response_and_survival_analysis__pfs",
+            "descriptor__response_and_survival_analysis__os",
+            "descriptor__response_and_survival_analysis__dfs",
+            "descriptor__response_and_survival_analysis__ttp",
```
Comment (gregorjerse) [inline resolwe_bio/filters.py:84]: I believe you should also add corresponding indices in the database, otherwise, these queries will be slow.

See https://github.com/genialis/resolwe-bio/blob/master/resolwe_bio/migrations/resolwe_indexes_2.sql for example how to add missing indices to JSON fields. 

Check also that the added indices are really applied to the query.

File: tests/urls.py [hunk 2]
```diff
@@ -3,15 +3,16 @@
...
+from resolwe.flow.views import CollectionViewSet, EntityViewSet
+from resolwe_bio.filters import BioCollectionFilter, BioEntityFilter
 from resolwe_bio.kb.views import (
     FeatureViewSet, MappingSearchViewSet
 )
 
 from .routers import SearchRouter
 
 
+CollectionViewSet.filter_class = BioCollectionFilter
```
Comment (gregorjerse) [inline tests/urls.py:15]: Maybe instead of patching URLs, one could use the `Composer` class. Otherwise, the URLs in the genialis-base repository will also have to be patched.

See https://github.com/genialis/resolwe/blob/90e4eb1160ac3fe0c2f0ac1b6bd18700cd82da55/resolwe/composer.py#L8 

There is some guidance on how to use the composer in

https://github.com/genialis/resolwe/blob/90e4eb1160ac3fe0c2f0ac1b6bd18700cd82da55/resolwe/flow/filters.py#L189

I believe you have to extend the filter used in the `CollectionViewSet` class, which is `resolwe.flow.filters.CollectionFilter`. 

Please try it out and remove line above if it is working.

**Review by gregorjerse — CHANGES_REQUESTED** [review]

File: resolwe_bio/extensions.py [hunk 3]
```diff
@@ -1,9 +1,65 @@
...
+from django.db.models import Q
 from django_filters import rest_framework as filters
 
 from resolwe.composer import composer
 
 
+class ExtendedCollectionFilter:
+    """Collection filter extension."""
+
+    descriptor__general__species = filters.CharFilter(
```
Comment (gregorjerse) [inline resolwe_bio/extensions.py:11]: You have to return collections that contain at least one sample (instance of Entity class) that has a descriptor with the given species.

So something like this:

```
descriptor__general__species = filters.CharFilter(
        field_name="entity__descriptor__general__species", lookup_expr="contains", distinct=True
    )
```

The distinct is necessary or collection with multiple samples matching given criteria will be returned multiple times.


All other queries also have to be modified.

File: resolwe_bio/migrations/resolwe_indexes_3.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
+-- Index for Django filter descriptor__general__species__icontains
+CREATE INDEX idx_collection_descriptor__general__species ON flow_collection USING gin (UPPER(descriptor #>> '{general,species}'::text[]) gin_trgm_ops );
```
Comment (gregorjerse) [inline resolwe_bio/migrations/resolwe_indexes_3.sql:2]: All indices have to be applied to the `flow_entity` table. Fix index names accordingly.

File: resolwe_bio/tests/unit/test_api.py [hunk 5]
```diff
@@ -30,6 +31,87 @@ def _check_filter(
...
+    @classmethod
+    def setUpTestData(cls):
+        cls.contributor = get_user_model().objects.create_user(
+            username="joemiller",
+            email="contributor@test.com",
+            first_name="Joe",
+            last_name="Miller",
+        )
+
+        cls.descriptor_schema = DescriptorSchema.objects.create(
```
Comment (gregorjerse) [inline resolwe_bio/tests/unit/test_api.py:53]: Fix tests: multiple collections containing multiple samples ...

**Review by gregorjerse — CHANGES_REQUESTED** [review]

File: docs/CHANGELOG.rst [hunk 6]
```diff
@@ -64,6 +64,7 @@ Fixed
 Added
 -----
 - Add bcftools version 1.14 to ``resolwebio/common`` Docker image
+- Add filters for species, tissue, treatment and outcome on BioCollection
```
Comment (gregorjerse) [inline docs/CHANGELOG.rst:67]: I would add something more descriptive and avoid using non-existing term BioCollection. Maybe something like 'Add ExtendedCollectionFilter filter to allow filtering collections by samples containing given species, tissue ...'.

**Review by gregorjerse — CHANGES_REQUESTED** [review]
Code now looks great. 
Please fix the commit message, rebase and then it can be merged.

**Review by gregorjerse — APPROVED** [review]
